firebase-firestore/src/test/java/com/google/firebase/firestore/model/MutationSquashTest.java

```diff
@@ -0,0 +1,250 @@
+
+    int caseNumber = 0;
+    caseNumber += runPermutationTests(docs, Lists.newArrayList(mutations));
+
+    // There are 4! * 3 cases
+    assertEquals(72, caseNumber);
+  }
+
+  @Test
+  public void testSquashMutationByCombinationsAndPermutations() {
+    List<MutableDocument> docs =
+        Lists.newArrayList(
+            doc("collection/key", 0, map("foo", "foo-value", "bar.baz", 1)),
+            deletedDoc("collection/key", 0),
+            unknownDoc("collection/key", 0));
+    List<Mutation> mutations =
+        Lists.newArrayList(
+            setMutation("collection/key", map("bar", "bar-value")),
+            setMutation("collection/key", map("bar.rab", "bar.rab-value")),
+            new DeleteMutation(key("collection/key"), Precondition.exists(true)),
+            patchMutation(
+                "collection/key",
+                map("foo", "foo-patched-value", "bar.baz", FieldValue.increment(1))),
+            patchMutation(
+                "collection/key", map("foo", "foo-patched-value", "bar.baz", FieldValue.delete())),
+            mergeMutation(
+                "collection/key",
+                map("arrays", FieldValue.arrayUnion(1, 2, 3)),
+                Arrays.asList(field("arrays"))));
+
+    int caseNumber = 0;
+    for (int subsetSize = 0; subsetSize <= mutations.size(); ++subsetSize) {
+      Set<Set<Mutation>> combinations = Sets.combinations(Sets.newHashSet(mutations), subsetSize);
+      for (Set<Mutation> combination : combinations) {
+        caseNumber += runPermutationTests(docs, Lists.newArrayList(combination));
+      }
+    }
+
+    // There are (0! + 6*1! + 15*2! + 20*3! + 15*4! + 6*5! + 6!) * 3 = 5871 cases.
+    assertEquals(5871, caseNumber);
+  }
+
+  private int runPermutationTests(List<MutableDocument> docs, List<Mutation> mutations) {
+    Timestamp now = Timestamp.now();
+    int caseNumber = 0;
+    List<List<Mutation>> permutations = generatePermutations(Lists.newArrayList(mutations));
+    for (MutableDocument doc : docs) {
+      for (List<Mutation> permutation : permutations) {
+        MutableDocument document = doc.clone();
+        MutableDocument docCopy = document.clone();
+        MutationSquash.Type type = MutationSquash.Type.None;
+        for (Mutation mutation : permutation) {
+          type = mutation.applyToLocalView(document, now, type);
+        }
+        Mutation squashed = MutationSquash.toMutation(docCopy, document, type);
+        if (squashed != null) {
+          squashed.applyToLocalView(docCopy, now, MutationSquash.Type.None);
+        }
+        assertEquals(getDescription(caseNumber, permutation, squashed), document, docCopy);
+
+        caseNumber += 1;
+      }
+    }
+    return caseNumber;
+  }
+
+  private String getDescription(int caseNumber, List<Mutation> mutations, Mutation squashed) {
+    StringBuilder builder = new StringBuilder();
+    builder.append("MutationSquash test (" + caseNumber + ") failed with:\n");
```

Comment:
FWIW I am not sure if the case number is helpful here (what does it mean if test 3342 failed?) The rest of the description should be sufficient to trace back to the original test.